source/extensions/clusters/aggregate/cluster_lb.h

```diff
@@ -0,0 +1,163 @@
+      context_ = owned_context_.get();
+    } else {
+      context_ = context;
+    }
+  }
+
+  // Upstream::LoadBalancerContext
+  absl::optional<uint64_t> computeHashKey() override { return context_->computeHashKey(); }
+  const Network::Connection* downstreamConnection() const override {
+    return context_->downstreamConnection();
+  }
+  const Router::MetadataMatchCriteria* metadataMatchCriteria() override {
+    return context_->metadataMatchCriteria();
+  }
+  const Http::HeaderMap* downstreamHeaders() const override {
+    return context_->downstreamHeaders();
+  }
+  const Upstream::HealthyAndDegradedLoad&
+  determinePriorityLoad(const Upstream::PrioritySet&,
+                        const Upstream::HealthyAndDegradedLoad& original_priority_load) override {
+    // Re-assign load. Set all traffic to the priority and availability selected in aggregate
+    // cluster.
+    // TODO(yxue): allow determinePriorityLoad to affect the load of top level cluster and verify it
+    // works with current retry plugin
+    size_t priorities = original_priority_load.healthy_priority_load_.get().size();
+    priority_load_.healthy_priority_load_.get().assign(priorities, 0);
+    priority_load_.degraded_priority_load_.get().assign(priorities, 0);
+
+    if (host_availability_ == Upstream::LoadBalancerBase::HostAvailability::Healthy) {
+      priority_load_.healthy_priority_load_.get()[host_priority_] = 100;
+    } else {
+      priority_load_.degraded_priority_load_.get()[host_priority_] = 100;
+    }
+    return priority_load_;
+  }
+  bool shouldSelectAnotherHost(const Upstream::Host& host) override {
+    return context_->shouldSelectAnotherHost(host);
+  }
+  uint32_t hostSelectionRetryCount() const override { return context_->hostSelectionRetryCount(); }
+  Network::Socket::OptionsSharedPtr upstreamSocketOptions() const override {
+    return context_->upstreamSocketOptions();
+  }
+
+private:
+  Upstream::HealthyAndDegradedLoad priority_load_;
+  std::unique_ptr<Upstream::LoadBalancerContext> owned_context_;
+  Upstream::LoadBalancerContext* context_{nullptr};
+  Upstream::LoadBalancerBase::HostAvailability host_availability_;
+  uint32_t host_priority_;
+};
+
+class AggregateClusterLoadBalancer : public Upstream::LoadBalancer,
+                                     Upstream::ClusterUpdateCallbacks {
+public:
+  AggregateClusterLoadBalancer(Upstream::ClusterManager& cluster_manager,
+                               const std::vector<std::string>& clusters,
+                               Upstream::ClusterStats& stats, Runtime::Loader& runtime,
+                               Runtime::RandomGenerator& random,
+                               const envoy::api::v2::Cluster::CommonLbConfig& common_config);
+
+  // Upstream::LoadBalancer
+  Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override {
+    // Initialize the inner load balancer.
+    if (!initialized_) {
```

Comment:
I'm not a huge fan of possibly doing a bunch of LB setup on the request path, is there somewhere earlier you can do this and just ASSERT that we're initialzied at this point? Maybe the ctor?

Reply:
I tried to put the initialization in the ctor but it will trigger segfault. The problem seems to be related with the initialization sequence of thread local cluster, thread aware load balancer and cluster. 

I also tried to put the initialization into the `initialize` of thread local aware balancer like refresh in `thread_aware_lb_impl.cc` but it doesn't work because we need to recreate the thread aware lb of the top level cluster rather than the inner cluster.

Reply:
I think we'll want to dig into this a bit more, having this inline initialization seems like a bit of a code smell. What do you mean by recreating the top level cluster vs inner cluster? Conceptually I'm imagining that you can initialize the LB with whatever clusters are available and then rely on the cluster update cbs to pick up new clusters that are added?

another idea: other clusters store the LB structures on the TLS cluster, not the LB. This means that the only thing that needs to care about cluster changes the cluster. maybe that makes the init story better?

Reply:
> Conceptually I'm imagining that you can initialize the LB with whatever clusters are available and then rely on the cluster update cbs to pick up new clusters that are added?

That's what I thought before but there are some hidden issues here related with the initialization sequence. I will dig it.

> another idea: other clusters store the LB structures on the TLS cluster, not the LB. This means that the only thing that needs to care about cluster changes the cluster. maybe that makes the init story better?

I tried to store the LB structure on the TLS cluster. But the TLS cluster interface doesn't provide function to update LB. The only way to update is triggered by the update of member change. For example, aggregate cluster `X` includes two clusters `A` and `B`. The host change of `A` and `B` will trigger the recreate of `A` and `B`'s LB only. We cannot rebuild the LB for `X`. We need some hook inside the cluster_manager_impl.

Reply:
why not just store all the state on the cluster and have the LB itself be stateless? it seems like you could store the priority set + the priority -> cluster mapping on the cluster and read from that in the LB? that way all updates are to the cluster and you dont need a mechanism to update the top level LB

Reply:
Create the stateless LB and update priority set in cluster.